### PR TITLE
Added an UnboundStablePriorityMailbox

### PR DESCRIPTION
**Pull request by AndreSteenbergen**
Added an UnboundStablePriorityMailbox, sending messages according to priority. Messages with the same priority will be send using the same order as they appear. #2652 is the related issue.

### CODE REVIEW

File: src/core/Akka/Dispatch/Mailbox.cs [hunk 1]
```diff
@@ -682,7 +682,7 @@ public override IMessageQueue Create(IActorRef owner, ActorSystem system)
     /// Priority mailbox base class; unbounded mailbox that allows for prioritization of its contents.
     /// Extend this class and implement the <see cref="PriorityGenerator"/> method with your own prioritization.
     /// The value returned by the <see cref="PriorityGenerator"/> method will be used to order the message in the mailbox.
-    /// Lower values will be delivered first. Messages ordered by the same number will remain in delivery order.
+    /// Lower values will be delivered first. Messages ordered by the same number will remain in delivered in undefined order.
```
Comment (Aaronontheweb) [inline src/core/Akka/Dispatch/Mailbox.cs]: Right

File: src/core/Akka/Dispatch/Mailbox.cs [hunk 2]
```diff
@@ -716,7 +716,45 @@ protected UnboundedPriorityMailbox(Settings settings, Config config) : base(sett
...
     }
 
-    //todo: bounded priority mailbox; stable priority mailboxes
+    //todo: bounded priority mailbox;
+
+    /// <summary>
+    /// Priority mailbox - an unbounded mailbox that allows for prioritization of its contents.
+    /// Extend this class and implement the <see cref="PriorityGenerator"/> method with your own prioritization.
+    /// The value returned by the <see cref="PriorityGenerator"/> method will be used to order the message in the mailbox.
+    /// Lower values will be delivered first. Messages ordered by the same number will remain in delivery order.
```
Comment (Aaronontheweb) [inline src/core/Akka/Dispatch/Mailbox.cs]: Looks good

File: src/core/Akka/Dispatch/Mailbox.cs [hunk 3]
```diff
@@ -716,7 +716,45 @@ protected UnboundedPriorityMailbox(Settings settings, Config config) : base(sett
...
+
+        /// <summary>
+        /// The default capacity of an unbounded priority mailbox.
+        /// </summary>
+        public const int DefaultCapacity = 11;
+
+        /// <inheritdoc cref="MailboxType"/>
+        public sealed override IMessageQueue Create(IActorRef owner, ActorSystem system)
+        {
+            return new UnboundedStablePriorityMessageQueue(PriorityGenerator, InitialCapacity);
```
Comment (Aaronontheweb) [inline src/core/Akka/Dispatch/Mailbox.cs]: LGTM

File: src/core/Akka/Util/StableListPriorityQueue.cs [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using Akka.Actor;
+
+namespace Akka.Util
+{
+    /// <summary>
+    /// Priority queue implemented using a simple list with binary search for inserts.
+    /// This specific implementation is cheap in terms of memory but weak in terms of performance.
```
Comment (Aaronontheweb) [inline src/core/Akka/Util/StableListPriorityQueue.cs]: Thanks for the comment, this is helpful.

**Comment by AndreSteenbergen** [comment]
I actually forgot I made this pull request. I will try to add unit tests asap. I do have some other priorities as well. If I am correct I am using this in production somewhere already ...

**Comment by Aaronontheweb** [comment]
@AndreSteenbergen all good - I added some myself. Had an urgent need for this feature and figured I could get it into a 1.3.13 release before I pull all of the v1.4 stuff into the `dev` branch.

**Comment by Aaronontheweb** [comment]
@AndreSteenbergen good news is that all of your code worked perfectly - threw a Property-based test at the queue structure itself to verify it and some tests on the mailbox too. All worked as expected.

**Comment by ismaelhamed** [comment]
> @AndreSteenbergen all good - I added some myself. Had an urgent need for this feature and figured I could get it into a 1.3.13 release before I pull all of the v1.4 stuff into the `dev` branch.

I ain't installing anything 1.3.13 in production...

**Comment by Aaronontheweb** [comment]
> I ain't installing anything 1.3.13 in production...

LOL :p - maybe I'll do what the airlines do and skip over having seat 13 and go straight to 14 instead :p

**Comment by AndreSteenbergen** [comment]
You can surely publish 1.3.13, just don't publish it on Friday the 13th ;)

**Pull request merged**